File: Java solutions/412. Fizz Buzz.java
class Solution {
    public List<String> fizzBuzz(int n) {
        // Naive Approach
        // time complexity: o(n)
        // space complexity: o(1)
        // List<String> answer = new ArrayList<>();
        // for (int i = 1; i <= n; i++){
        //     if (i % 3 == 0 && i % 5 == 0) answer.add("FizzBuzz");
        //     else if (i % 3 == 0) answer.add("Fizz");
        //     else if (i % 5 == 0) answer.add("Buzz");
        //     else answer.add(Integer.toString(i));
        // }
        // return answer;
        
        // String concatenation, neater solution if the game rule changes
        // time complexity: o(n)
        // space complexity: o(n)
        // List<String> answer = new ArrayList<>();
        // for (int i = 1; i <= n; i++){
        //     String cur = "";
        //     if (i % 3 == 0) cur += "Fizz";
        //     if (i % 5 == 0) cur += "Buzz";
        //     if (cur.equals("")) cur += Integer.toString(i);
        //     answer.add(cur);
        // }
        // return answer;
        
        // hash table, for dealing with too many mappings
        List<String> answer = new ArrayList<>();
        HashMap<Integer, String> hm = new HashMap<>();
        hm.put(3, "Fizz");
        hm.put(5, "Buzz");
        for (int i = 1; i <= n; i++){
            String cur = "";
            for (Integer key: hm.keySet()){
                if (i % key == 0) cur += hm.get(key);
            }
            if (cur.equals("")) cur += Integer.toString(i);
            answer.add(cur);
        }
        return answer;
    }
}

File: Python solutions/412. Fizz Buzz.py
class Solution:
    def fizzBuzz(self, n: int) -> List[str]:
        # naive approach
        # time complexity: o(n)
        # space complexity: o(n)
        # ans = []
        # for num in range(1, n+1):
        #     if num % 3 == 0 and num % 5 == 0: ans.append("FizzBuzz")
        #     elif num % 3 == 0: ans.append("Fizz")
        #     elif num % 5 == 0: ans.append("Buzz")
        #     else: ans.append(str(num))
        # return ans
        
        # string concatenation for neater code
        # ans = []
        # for num in range(1, n+1):
        #     s = ""
        #     if num % 3 == 0: s += "Fizz"
        #     if num % 5 == 0: s += "Buzz"
        #     if s == "": s = str(num)
        #     ans.append(s)
        # return ans
        
        # hashmap if there's too many mapping
        ans = []
        hm = {3:"Fizz", 5:"Buzz"}
        for num in range(1, n+1):
            s = ""
            for key in hm.keys():
                if num % key == 0: s += hm[key]
            if s == "": s += str(num)
            ans.append(s)
        return ans
